Add acceptance tests for `secret` commands

andyfeller opened the pull request:
This PR expands upon the GitHub CLI acceptance tests to cover `gh secret` commands for 1) repository, 2) repository environment, and 3) organization secrets.

```shell
set -o pipefail && GH_ACCEPTANCE_TOKEN=$(gh auth token) GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=tinyfists go test -tags acceptance -json -run ^TestSecrets$ github.com/cli/cli/v2/acceptance  | tparse --all go test
```

resulting in:

```shell
┌─────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │            TEST             │             PACKAGE               │
│─────────┼─────────┼─────────────────────────────┼───────────────────────────────────│
│  PASS   │    4.82 │ TestSecrets/secret-repo-env │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    4.39 │ TestSecrets/secret-repo     │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    1.80 │ TestSecrets/secret-org      │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestSecrets                 │ github.com/cli/cli/v2/acceptance  │
└─────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │  5.35s  │ github.com/cli/cli/v2/acceptance │  --   │  4   │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

### Additional notes

- GitHub Actions uppercases all secret names
- Organization secrets require `admin:org` scope on token
- Repository environments are only available to paid GitHub plans, require running `secret-repo-env` against a different organization than other testscripts

andyfeller committed:
Add testscripts for `gh secret`, helper cmds

These are initial testscripts for confidence around core `gh secret` subcommands around 1) repository, 2) repository environment, 3) organization secrets.

Each testscript exercises the 3 core subcommands of `gh secret`: 1) setting, 2) listing, and 3) deleting each type of secret.

Note should be made that repository environments do not exist for free accounts such as `gh-acceptance-testing`.

andyfeller committed:
Update secret note about potential failure

Environments are only supported for GitHub Actions with certain plans; `gh-acceptance-testing` at this time does not have this support.

andyfeller committed:
Refactor env2upper, env2lower; add docs

This commit refactors 2 new custom commands to work more like the native env command, allowing for multiple environment variables to be set.

Additionally, the documentation on acceptance tests have been expanded to list out custom commands.

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

andyfeller committed:
Fix typo in custom command doc

williammartin commented on acceptance/testdata/secret/secret-org.txtar:
```diff
@@ -0,0 +1,16 @@
+# Prepare organization secret, GitHub Actions uppercases secret names
+env2upper ORG_SECRET_NAME=$RANDOM_STRING
```
**Change Requested**

Please include the `$SCRIPT_NAME` in any globally visible resources so that if they are left lying around we have an indicator of where they came from.

If we can't hyphens for some reason then let's santitize SCRIPT_NAME before setting the env var.

andyfeller replied:
Makes sense though I assume you aren't opposed to having the random string being used as well in the secret name.  I ask because we don't have the convenience of a repository to isolate testing from different people, causing a static name to potentially conflict if multiple people are running these.

williammartin replied:
Yep, I still think we should be using the `$RANDOM_STRING` as part of the secret name.

williammartin commented on acceptance/testdata/secret/secret-org.txtar:
```diff
@@ -0,0 +1,16 @@
+env2upper ORG_SECRET_NAME=$RANDOM_STRING
+
+# Confirm organization secret does not exist, will fail admin:org scope missing
+! exec gh api /orgs/$ORG/actions/secrets/$ORG_SECRET_NAME
```
**Question**

Can you talk to me a bit about why you feel you need to check a secret with this name doesn't exist before proceeding? How's this moving the needle on your confidence levels?

**Suggestion**

If you feel really strongly about having this precondition then I would propose using:

```
exec gh secret list --org $ORG
! stdout $ORG_SECRET_NAME
```

That way it's easy to see the duality of the before and after the secret is set.

andyfeller replied:
> Can you talk to me a bit about why you feel you need to check a secret with this name doesn't exist before proceeding? How's this moving the needle on your confidence levels?

I had 2 concerns initial writing this test:

1. Organization resources don't have a neat container like a repository that can be cleaned up when test completes
2. The permissions needed for working with organization secrets and variables might be uncommon to token used for testing

So I thought confirming that the random string I'm about to use for testing didn't exist as a way to ensure no prior state would conflict and the token has sufficient privileges before moving on.

jtmcg replied:
Interesting... I am doing something similar in my PR. I've dropped a [comment](https://github.com/cli/cli/pull/9783/files#r1805452914) there to have a tightly scoped discussion on it, because ultimately I think @andyfeller's reasoning here is sound but isn't relevant to my case, which still falls within the scope of @williammartin's question.

williammartin replied:
Thanks! I responded over on Tyler's PR as well about why I think a pre-assertion could be valuable in my eyes.

> Organization resources don't have a neat container like a repository that can be cleaned up when test completes

It feels very similar to `repo create` commands, where we don't check that a repo with a `$RANDOM_STRING` name doesn't exist before creating it.

> The permissions needed for working with organization secrets and variables might be uncommon to token used for testing

That's an interesting point. I have ideas around making scopes requirements clear that would remove the need for this.

How would you feel if each script started with a command like:

```
scopes admin:org,repo,... 
```

Such that each script enumerated the scopes required to continue with the test? It feels like it would be very informative to the reader.

williammartin commented on acceptance/testdata/secret/secret-org.txtar:
```diff
@@ -0,0 +1,16 @@
+exec gh secret set $ORG_SECRET_NAME --org $ORG --body 'does not matter as cannot read it once set'
+
+# Defer organization secret cleanup
+defer gh secret delete $ORG_SECRET_NAME --org $ORG
```
Love it. I totally forgot this would be necessary as I was reading it.

williammartin commented on acceptance/testdata/secret/secret-repo-env.txtar:
```diff
@@ -0,0 +1,28 @@
+# Force GitHub CLI to treat testscript as TTY
+env GH_FORCE_TTY=80
```
**Question**

Can you talk to me a bit about why you feel you need to have a TTY to have confidence that repo secrets are being interacted with correctly?

**Food For Thought**

So with the current focus on verifying the APIs are working correctly I think at the moment I'm generally in favour of only asserting stdio when we're verifying the expected state of resources after acting, rather than generally checking `gh` stdio looks a certain way. 

Put another way, we're not writing a general set of tests for `gh` UX here, and if we were there's a lot more work that needs to be done to support that. I'd rather that we focus on our specific goal and then later we can holistically tackle the problem of stdio.

I really don't want to end up in a place where we all have different ideas on what stdio we should be asserting.

**Suggestion**

If we do want a TTY, I guess I'd prefer `GH_FORCE_TTY=true` because setting the value to a specific width makes me think the width is important (and I don't think it is?)

andyfeller replied:
> Can you talk to me a bit about why you feel you need to have a TTY to have confidence that repo secrets are being interacted with correctly?

The results of `gh secret` appeared to be acting in a non-TTY manner, which resulted in certain commands having no output in certain situations, making it difficult to test.

> Put another way, we're not writing a general set of tests for gh UX here, and if we were there's a lot more work that needs to be done to support that. I'd rather that we focus on our specific goal and then later we can holistically tackle the problem of stdio.

Could you help me understand how this test doesn't satisfy requirements or what goal you're referring to?

> Given I set a variable with gh variable set
> When I run gh variable list
> Then It is included in the list
> And When I run gh variable get targeting that variable
> Then I see the value that I set

williammartin replied:
> The results of gh secret appeared to be acting in a non-TTY manner, which resulted in certain commands having no output in certain situations, making it difficult to test.

What's not clear to me is **why** you need to test the output of these commands. What confidence about `gh` working with tenants does checking stdout here give you:

```
exec gh secret set TESTSCRIPTS_ENV --env testscripts --body 'does not matter as cannot read it once set'
stdout 'Set Actions secret TESTSCRIPTS_ENV for'
```

That this doesn't:

```
exec gh secret list --env testscripts
stdout 'TESTSCRIPTS_ENV\s+less than a minute ago'
```

> Could you help me understand how this test doesn't satisfy requirements

It absolutely satisfies the Acceptance Criteria. 

What it also does is introduce a difference in how the scripts are being written because it:
 * Asserts on things that other scripts aren't
 * Sets `gh` up to use TTY in a way that other scripts aren't

Repeating what I wrote in my [comment to Tyler](https://github.com/cli/cli/pull/9783/files#r1806157923):

> But I want to be clear here, I'm not saying that these should be removed. I am saying that I think it's important that there is a clear and shared pattern to all the testscripts that we can say "make it look like that". That can also mean going back to the scripts that are already being merged and making them use this pre and post pattern.

I might be happy with "do whatever you need to do to assert stdout of every exec", I might be happy with "only assert on stdout of non-TTY", and I might be happy with "only assert on stdout of commands that are being used to verify state rather than cause mutations" but I'm not really happy with "all the tests do different things depending on who wrote them" because the knock on effects of that have the potential to be kind of annoying.

So I'm trying to understand how valuable these assertions are to you and to the script itself, and to trade that off against any costs, so that we can tease out the guidelines we want to encourage.

Hope that makes sense.

andyfeller replied:
> So I'm trying to understand how valuable these assertions are to you and to the script itself, and to trade that off against any costs, so that we can tease out the guidelines we want to encourage.

I can acquiesce that some of these assertions might not be necessary, or rather that some might not find them necessary.

Personally, I value asserting the context of commands as closed to where they are being made as possible.  It isn't always doable hence sometimes I assert the output of the command while later on trying to use the information and assert that, too.   You could say I care about how the product behaves as well as the state it is changing.  Again, I can acquiesce my personal values for now as I have refactored these tests in the manner of `gh workflow` testscripts, which I feel increases my confidence more.

BagToad replied:
>  but I'm not really happy with "all the tests do different things depending on who wrote them" because the knock on effects of that have the potential to be kind of annoying.

I really agree with this. 

:2c: When I went into writing some of these tests, I was under the impression that interactivity is out of scope. Forcing TTY on feels like we risk stepping over that line, or perhaps risking a command accidentally prompting and hanging the tests for some reason.

I also think that asserting that output looks a certain way is something already covered by our unit tests. 

I'm wondering what asserting on forced TTY output gains us over the assertion that comes with `exec` & the exit code?

williammartin commented on acceptance/testdata/secret/secret-repo-env.txtar:
```diff
@@ -0,0 +1,28 @@
+cd $SCRIPT_NAME-$RANDOM_STRING
+
+# Create a repository environment, will fail if organization does not have environment support
+exec gh api /repos/$ORG/$SCRIPT_NAME-$RANDOM_STRING/environments/testscripts -X PUT --jq '.name'
```
**Question**

Do we need to manually create a `testscripts` environment in whatever org is being targeted? I might misunderstand how the API works for secret setting, I assumed the environment would have to exist already.

andyfeller replied:
These are repository environments, meaning they are unique to the repository themselves, so it isn't possible to assume this is setup within the organization in advance.

williammartin replied:
Oh durrrr. Very sorry. I absolutely glossed over your very clear comment here and didn't read the `PUT` at all. I understand now.

williammartin commented on acceptance/testdata/secret/secret-repo-env.txtar:
```diff
@@ -0,0 +1,28 @@
+
+# Verify no repository environment secrets exist
+exec gh secret list --env testscripts
+stderr 'no secrets found'
```
**Suggestion**

Same suggestion as the org secrets to make the pre and post assertions look similar.

```
exec gh secret list --env testscripts
! stdout 'TESTSCRIPTS_ENV\s+less than a minute ago'
```

williammartin commented on acceptance/testdata/secret/secret-repo-env.txtar:
```diff
@@ -0,0 +1,28 @@
+exec gh repo clone $ORG/$SCRIPT_NAME-$RANDOM_STRING
+cd $SCRIPT_NAME-$RANDOM_STRING
+
+# Create a repository environment, will fail if organization does not have environment support
```
**Question**

Do you think this test should run at all if the org doesn't have environment support? We could `cond` the whole test script so that it skips if there isn't support.

andyfeller replied:
That is a really good question!  We could skip this test, but I somewhat would prefer we have an environment where we don't have to worry about that and ensuring that core commands work as expected.

williammartin replied:
Yeh so Kynan suggested we just make this work for our org, and that seems like an obvious first step. I'm not sure how that works on ghe.com, need to investigate. Beyond that I can imagine that if anyone else were to run these tests we might want to add `cond` support.

This is not something I would prioritise until other people running them was an outcome we wanted to achieve, just looking for your general opinions.

andyfeller replied:
Unfortunately I think the only way this could happen is attempting to pull plan information from the API about the organization or user account being targeted from

- [Get an organization](https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#get-an-organization)
- [Get the authenticated user](https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28#get-the-authenticated-user)

williammartin commented on acceptance/README.md:
````diff
@@ -61,6 +61,39 @@ The following custom environment variables are made available to the scripts:
+  defer gh repo delete --yes $ORG/$SCRIPT_NAME-$RANDOM_STRING
+  ```
+
+- `env2lower`: set environment variable to the lowercase version of another environment variable
````
**Question**

Is this used anywhere? Is there somewhere we expect it to be used?

If not...

**Change Requested**

Please remove it.

andyfeller replied:
I assume this request is because you value only having code around that is actively used.  I value having tools readily at hand for various scenarios.  One such scenario I can foresee but haven't written the test to require is asserting the repository name, which is always lowercase.

If there was a way to handle dynamic, case insensitive comparison where this wasn't needed, I'd get rid of both custom commands all together.

jtmcg replied:
_Subscribe_ 😄 

(I'm interested in this discussion and want to see where it goes, so jumping on the thread via this comment so I'll get notifications about it)

williammartin replied:
> I assume this request is because you value only having code around that is actively used.

Actively used or at least with the intent to use, which is why I asked whether you expected it to be used somewhere.

Butttt it's a little more nuanced than whether the code is actively used or not.

Mostly I'm interested in avoiding a proliferation of commands. With each new command we add, the larger potential variation in the tests we have to maintain. I think new commands should have a fairly high bar for inclusion and should be supported by their usage. I'm keen to avoid a situation where commands are added because they might be useful, and then because they exist they get used.

All that said, if you have a plan of where it's going to be used and you don't want to wait until then to include it, I'm not going to use any more virtual ink on the subject.

williammartin commented on acceptance/testdata/secret/secret-org.txtar:
```diff
@@ -0,0 +1,16 @@
+# Prepare organization secret, GitHub Actions uppercases secret names
+env2upper ORG_SECRET_NAME=$RANDOM_STRING
```
**Food For Thought**

If this is the only place that casing matters, maybe we can inject an already upper cased env var so we don't increase the command surface area? No need to rush a decision on it though, happy to revisit as we understand usage more.

andyfeller replied:
This will matter in `gh variable`, too.  And technically there are places where certain things should be all lowercase.

williammartin replied:
Sorry, I should have been clearer.

> If this is the only place that casing matters

to

> If upper or lowercasing an environment variable that is injected from Go is the only place that casing matters

Thus if we had `$RANDOM_STRING` (or `$RANDOM_STRING_L` /  `$RANDOM_STRING_LOWER and `$RANDOM_STRING_U` / `$RANDOM_STRING_UPPER`) then the commands go away.

But maybe we also would need to do the same for `$SCRIPT_NAME`, and that seems like it starts to get annoying.

Let's leave it as is.

williammartin commented on acceptance/README.md:
```diff
@@ -61,6 +61,39 @@ The following custom environment variables are made available to the scripts:
  * `HOME`: Set to the initial working directory. Required for `git` operations
  * `GH_CONFIG_DIR`: Set to the initial working directory. Required for `gh` operations
 
+#### Custom Commands
```
Interesting idea!

williammartin reviewed (changes requested):
Firstly, please don't take my many comments as a criticism, there are only 1 or possibly 2 requested changes here. 

Really, the reason there are so many comments is because you've managed to pick a command with a lot of interesting quirks and because you've brought a different perspective and taken some things in a different direction than the scripts @BagToad or I wrote so far. We just need to work through these to make sure we end up with a consistent approach to tests, even if that means we go back to adjust the already written ones.

Thanks!

andyfeller committed:
Minor polish / consistency

andyfeller committed:
Merge branch 'trunk' into andyfeller/testscripts-secret

andyfeller committed:
Refactor `gh secret` testscript

This is a bit of a refactor based on the work done in `gh workflow` as a better approach to verify secrets created are what we expect.

Changes made:

1. Removed `env2lower` as it wasn't being used in testscripts
2. Added `replace` custom command to deal with testing organization workflow secrets
3. Refactored secret testscripts to create and run workflow that tests the value of the secret provided
4. Minor reordering of test `acceptance` test functions as appending to the end is confusing and adds conflicts
5. Removed stdout TTY assertions

andyfeller commented on acceptance/testdata/secret/secret-org.txtar:
```diff
@@ -0,0 +1,85 @@
+    steps:
+      - name: Assert secret value matches
+        env:
+          ORG_SECRET: ${{ secrets.$SECRET_NAME }}
```
I've tried my best in the custom command definition to explain why `replace` command is necessary here.  I'm all ears for ideas to do this more simply, however there's nothing native to testscripts to handle dynamic aspects like this.

williammartin commented on acceptance/testdata/secret/secret-org.txtar:
```diff
@@ -0,0 +1,85 @@
+# Setup environment variables used for testscript
+env REPO=$SCRIPT_NAME-$RANDOM_STRING
+env2upper SECRET_NAME=$SCRIPT_NAME_$RANDOM_STRING
```
This isn't expanding like you'd expect because `os.Expand` is being given `SCRIPT_NAME_` and `RANDOM_STRING`.

You'd need to do:

```
env2upper SECRET_NAME=${SCRIPT_NAME}_${RANDOM_STRING}
```

However, even if it were... we'd still have an issue with hyphens in the file names.

I suggest that we adjust `SCRIPT_NAME` env var like so:

```
ts.Setenv("SCRIPT_NAME", strings.ReplaceAll(scriptName, "-", "_"))
```

I don't see any particular reason to favour hyphens. If we find a reason in the future we can come up with another solution.

To be honest, it might be worth a follow up PR where we change all env vars to expand with `{}` and add a guideline.

andyfeller replied:
> This isn't expanding like you'd expect because `os.Expand` is being given `SCRIPT_NAME_` and `RANDOM_STRING`.
> 
> You'd need to do:
> 
> ```
> env2upper SECRET_NAME=${SCRIPT_NAME}_${RANDOM_STRING}
> ```

> To be honest, it might be worth a follow up PR where we change all env vars to expand with {} and add a guideline.

Makes sense and aligns with common Shell scripting practices

williammartin commented on acceptance/acceptance_test.go:
```diff
@@ -146,6 +155,58 @@ func sharedCmds(tsEnv testScriptEnv) map[string]func(ts *testscript.TestScript,
+			ts.Logf("replace src: %s", src)
+			info, err := os.Stat(src)
+			ts.Check(err)
+			mode := info.Mode() & 0o777
```
I think the idea here is that we always write the file with the same permissions it already had? If so, maybe you can write a comment explaining that.

williammartin commented on acceptance/testdata/secret/secret-repo.txtar:
```diff
@@ -0,0 +1,76 @@
+cd $REPO
+
+# Create a repository secret
+exec gh secret set TESTSCRIPTS --body 'just a repository secret'
```
I like that these don't do the variable business because they are already scoped. Nice.

williammartin commented on acceptance/acceptance_test.go:
```diff
@@ -27,13 +27,13 @@ func TestMain(m *testing.M) {
 	}))
 }
 
-func TestPullRequests(t *testing.T) {
```
It's a shame about this diff because it's going to make it kinda janky to track back the git history. It's not the end of the world though. Maybe once we're happy with this PR we can squash it all down and only have the addition of `TestSecrets` in this part.

andyfeller replied:
Aside from being able to skim the list of tests, the conflict with these tests being added at the end of the ever growing list felt like it could benefit from being alphabetized 🤷

williammartin replied:
Ah, from the diff it just looked like things had been bounced around by accident. Sure, is fine with me.

andyfeller replied:
For now, I'm just going to undo my reorganizing changes around the test functions.  I don't like it but want to get this done despite value in being able to scan the test functions alphabetically.

williammartin replied:
That's fine. I would accept a follow up PR to move them around given your reasoning. Makes sense to me.

williammartin reviewed (changes requested):
Pretty much there but one thing I do think needs to be resolved in the env var expansion for `secret-org`.

andyfeller committed:
Address @williammartin PR feedback

williammartin reviewed (approved):
LGTM once the conflict is resolved. We may want to follow up with some kind of conditional behaviour for the extra org permissions but these are not currently auto-running anywhere so we have the knowledge of what should and shouldn't work.

Thanks a lot for this, I know it wasn't the smoothest sailing.

andyfeller committed:
Merge branch 'trunk' into andyfeller/testscripts-secret

Mjdesignco commented:
@andyfeller
